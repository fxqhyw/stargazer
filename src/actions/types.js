export const GET_REPOSITORY = 'GET_REPOSITORY'
export const GET_REPOSITORY_SUCCESSFUL = 'GET_REPOSITORY_SUCCESSFUL'
export const GET_REPOSITORY_FAILURE = 'GET_REPOSITORY_FAILURE'
export const REMOVE_REPOSITORY = 'REMOVE_REPOSITORY'