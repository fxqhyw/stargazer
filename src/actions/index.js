import * as types from './types';

export const getRepository = (fullName) => ({
  type: types.GET_REPOSITORY,
  payload: fullName 
})

export const getRepositorySuccessful = (repository) => ({
  type: types.GET_REPOSITORY_SUCCESSFUL,
  payload: repository
})

export const getRepositoryFailure = (error) => ({
  type: types.GET_REPOSITORY_FAILURE,
  payload: error
})

export const removeRepository = (id) => ({
  type: types.REMOVE_REPOSITORY,
  payload: id
})

export const getRepositoryLanguages = (id) => ({
  type: types.GET_REPOSITORY_LANGUAGES,
  payload: id
})

export const getRepositoryLanguagesSuccessful = ({id, languages}) => ({
  type: types.GET_REPOSITORY_LANGUAGES_SUCCESSFUL,
  payload: {
    id: id,
    languages: languages
  }
})
