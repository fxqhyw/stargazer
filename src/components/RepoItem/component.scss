tr.repo-item {
  cursor: pointer
}

tr.repo-item:hover {
  color: seagreen;
}
