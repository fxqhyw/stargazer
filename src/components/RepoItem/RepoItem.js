import React from 'react'
import { useDispatch } from 'react-redux'
import { removeRepository } from '../../actions'
import { useHistory } from 'react-router-dom'
import './RepoItem.css'

const RepoItem = ({id, name, starsCount}) => {
  let history = useHistory()
  const dispatch = useDispatch()

  const handleRemove = () => {
    dispatch(removeRepository(id))
  }
 
  return(
    <tr
      className='repo-item'
      onClick={() => history.push(`/repositories/${id}`)}>
      <td>{name}</td>
      <td>{starsCount}</td>
      <td>
        <button
          type="button"
          className="close"
          onClick={handleRemove}>
          <span aria-hidden="true">&times;</span>
        </button>
      </td>
    </tr>
  )
}

export default RepoItem