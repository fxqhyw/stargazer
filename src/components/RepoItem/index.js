import RepoItem from './component'

export default RepoItem
