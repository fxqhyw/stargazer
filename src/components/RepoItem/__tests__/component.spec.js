import React from 'react'
import { useDispatch } from 'react-redux'
import { mount, shallow } from 'enzyme'
import RepoItem from '../component'
import { removeRepository } from '../../../actions'

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  })
}))

jest.mock('react-redux', () => ({
  useDispatch: jest.fn()
}))
 
describe('RepoItem', () => {
  let component;
  const repository = {
    id: 123,
    full_name: "facebook/react",
    stargazers_count: 12345,
    subscribers_count: 6678,
    forks_count: 30270,
    clone_url: "https://github.com/facebook/react.git",
    languages_url: "https://api.github.com/repos/facebook/react/languages"
  }
  
  it('should render correctly', () => {
    component = shallow(
      <RepoItem {...repository} />
    )
    expect(component).toMatchSnapshot()
  });
 
  it('should dispatch an action on button click', () => {
    const dispatch = jest.fn()
    useDispatch.mockImplementation(() => dispatch)

    component = mount(
      <table>
        <tbody>
          <RepoItem {...repository} />
        </tbody>
      </table>
    )
    component.find('button.close').simulate('click')
    expect(dispatch).toHaveBeenCalledTimes(1)
    expect(dispatch).toHaveBeenCalledWith(removeRepository(123))
  });

  afterEach(() => {
    component.unmount()
  })
});
