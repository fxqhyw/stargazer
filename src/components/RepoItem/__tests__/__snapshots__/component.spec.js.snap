// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoItem should render correctly 1`] = `
<tr
  className="repo-item"
  onClick={[Function]}
>
  <td />
  <td />
  <td>
    <button
      className="close"
      onClick={[Function]}
      type="button"
    >
      <span
        aria-hidden="true"
      >
        Ã—
      </span>
    </button>
  </td>
</tr>
`;
